backend/ota_demo_api/repository/search_repository.py

```diff
@@ -19,150 +87,166 @@ async def fetch(self, search_data: SearchRequest) -> Optional[List[ClusterSearch
+                        ROW_NUMBER() OVER (
+                            PARTITION BY ty_id ORDER BY (language != 'all', trip_type != 'all') DESC
+                        ) AS rn
+                    FROM cluster_search
+            """
 
         if is_city_country:
             query +=  """
                 WHERE lower(city) = lower(:city) AND lower(country) = lower(:country)
             """
-            query_params["city"] = search_data.city
-            query_params["country"] = search_data.country
 
         if is_coordinates:
             query += """
                 WHERE earth_box(ll_to_earth (:lat, :long), :radius) @> ll_to_earth (latitude, longitude)
                 AND earth_distance(ll_to_earth (:lat, :long), ll_to_earth (latitude, longitude)) < :radius
             """
-            query_params["lat"] = search_data.lat
-            query_params["long"] = search_data.long
-            query_params["radius"] = search_data.radius
 
         if search_data.trip_type:
             query += """
                 AND trip_type IN (:trip_type, 'all')
             """
-            query_params["trip_type"] = search_data.trip_type
         else:
             query += """
                 AND trip_type = :trip_type
             """
-            query_params["trip_type"] = "all"
 
         if search_data.language:
             query += """
                 AND language IN (:language, 'all')
             """
-            query_params["language"] = search_data.language
         else:
             query += """
                 AND language = :language
             """
-            query_params["language"] = "all"
 
         query += """
                 AND datapoint = ANY(:data_points)
                 GROUP BY ty_id, trip_type, language
             )
-            SELECT 
-                  sr.ty_id,
-                  sr.data_points,
-                  sr.scores, 
-                  sr.review_counts,
-                  sr.language,
-                  sr.trip_type,
-                  sr.rn,
-                  sr.search_score * (
-                    (
-                        (CASE WHEN sr.language = :language THEN 1 ELSE 0 END) + 
-                        (CASE WHEN sr.trip_type = :trip_type THEN 1 ELSE 0 END) + 
-                        array_length(sr.data_points, 1)
-                    ) / (array_length(:data_points, 1)::decimal + 2)
-                  ) as match_score,
-                  cs.score as score
-            FROM search_results sr
-            JOIN public.cluster_search cs ON (sr.ty_id = cs.ty_id)
-            WHERE cs.datapoint = 'oall' and cs.trip_type = 'all' and cs.language = 'all' and sr.rn = 1
         """
 
-        data_points = []
-        if search_data.categories or search_data.hotel_types:
-            data_points += search_data.categories or []
-            data_points += search_data.hotel_types or []
+        if is_count_query:
+            query += """
+                SELECT COUNT(*)
```

Comment:
[minor] I suggest to define `select_column_list` for if/else
and then we can use `query += f"""SELECT {select_column_list} FROM search_results ...."""` to reuse all other query elements.

Reply:
Good idea, refacted.